python/test/test_hail/methods/test_statgen.py

```diff
@@ -1129,3 +1130,273 @@ def test_window_by_locus(self):
+                               1: hl.Struct(locus=hl.Locus('2', 1), alleles=['A', 'AT', '@']),
+                               2: hl.Struct(locus=hl.Locus('2', 1), alleles=['AC', 'GT'])})
+        mt = mt.annotate_rows(**variants[mt.row_idx]).partition_rows_by('locus', 'locus', 'alleles')
+        r = hl.summarize_variants(mt, show=False)
+        self.assertEqual(r.n_variants, 3)
+        self.assertEqual(r.contigs, {'1': 1, '2': 2})
+        self.assertEqual(r.allele_types, {'SNP': 2, 'MNP': 1, 'Unknown': 1, 'Insertion': 1})
+        self.assertEqual(r.allele_counts, {2: 1, 3: 2})
+
+    @staticmethod
+    def _standardize_cols(a):
+        a = a.copy()
+        col_means = np.mean(a, axis=0, keepdims=True)
+        a -= col_means
+        col_lengths = np.linalg.norm(a, axis=0, keepdims=True)
+        col_filter = col_lengths > 0
+        return np.copy(a[:, np.squeeze(col_filter)] / col_lengths[col_filter])
+
+    def test_linear_mixed_model_low_rank(self):
+        seed = 0
+        n_populations = 8
+        fst = n_populations * [.9]
+        n_samples = 500
+        n_variants = 200
+        n_orig_markers = 100
+        n_culprits = 10
+        n_covariates = 3
+        sigma_sq = 1
+        tau_sq = 1
+
+        from numpy.random import RandomState
+        prng = RandomState(seed)
+
+        x = np.hstack((np.ones(shape=(n_samples, 1)),
+                       prng.normal(size=(n_samples, n_covariates - 1))))
+
+        mt = hl.balding_nichols_model(n_populations=n_populations,
+                                      n_samples=n_samples,
+                                      n_variants=n_variants,
+                                      fst=fst,
+                                      seed=seed)
+
+        pa_t_path = utils.new_temp_file(suffix='bm')
+        a_t_path = utils.new_temp_file(suffix='bm')
+
+        BlockMatrix.write_from_entry_expr(mt.GT.n_alt_alleles(), a_t_path)
+
+        a = BlockMatrix.read(a_t_path).T.to_numpy()
+        g = a[:,-n_orig_markers:]
+        g_std = self._standardize_cols(g)  # filters constant cols
+
+        n_markers = g_std.shape[1]
+
+        k = (g_std @ g_std.T) * n_samples / n_markers
+
+        beta = np.arange(n_covariates)
+        beta_stars = np.array([1] * n_culprits)
+
+        y = prng.multivariate_normal(
+            np.hstack((a[:, 0:n_culprits], x)) @ np.hstack((beta_stars, beta)),
+            sigma_sq * k + tau_sq * np.eye(n_samples))
+
+        # low rank computation of S, P
+        l = g_std.T @ g_std
+        sl, v = np.linalg.eigh(l)
+        n_eigenvectors = int((sl > 1e-10).sum())
+        sl = sl[-n_eigenvectors:]
+        v = v[:, -n_eigenvectors:]
+        s = sl * (n_samples / n_markers)
+        p = (g_std @ (v / np.sqrt(sl))).T
+
+        # compare with full rank S, P
+        sk0, uk = np.linalg.eigh(k)
+        sk = sk0[-n_eigenvectors:]
+        pk = uk[:, -n_eigenvectors:].T
+        assert np.allclose(sk, s)
```

Comment:
you're not using `self.assertTrue`, but I think this is a good thing -- pytest overrides Python assertions to produce nice error messages, so we can do this more.